util/reggen/fpv_csr.sv.tpl

```diff
@@ -113,160 +118,204 @@ module ${block.name}_csr_assert_fpv import tlul_pkg::*; (
         (d2h.d_error || d2h.d_data == 0);
   endproperty
 
+  // TODO: currently not used, will use once support regwen reg
   property wr_regen_stable_P(bit regen, bit [DWidth-1:0] exp_data);
     (!regen && $stable(regen)) |-> $stable(exp_data);
   endproperty
 
-// for all the regsters, declare assertion
 % for r in block.regs:
 <%
-  i = -1
-  msb = -1
-  lsb = 0
+  has_q = r.get_n_bits(["q"]) > 0
+  has_d = r.get_n_bits(["d"]) > 0
+%>\
+  % if not r.get_field_flat(0).shadowed:
+  % if r.is_multi_reg():
+<%
+      mreg_name = r.name
+      mreg_width_list = list()
+      mreg_fpv_name_list = list()
+      mreg_dut_path_list = list()
+      mreg_num_regs = r.get_n_fields_flat()
+
+      mreg_msb = -1
+      mreg_lsb = 0
+      i = 0
+%>\
+   % if not r.ishomog:
+     % for field in r.get_reg_flat(0).fields:
+<%
+  mreg_fpv_name_list.append(mreg_name + "_" + field.get_basename())
+  mreg_dut_path_list.append(mreg_name + "[s]." + field.get_basename())
+  mreg_width_list.append(field.msb - field.lsb + 1)
+%>\
+     % endfor
+<%
+  mreg_num_base_fields = len(mreg_fpv_name_list)
+  mreg_num_regs = mreg_num_regs / mreg_num_base_fields
+%>\
+   % else:
+<%
+  f = r.get_field_flat(0)
+  mreg_num_base_fields = 1
+  mreg_fpv_name_list.append(mreg_name)
+  mreg_dut_path_list.append(mreg_name + "[s]")
+  mreg_width_list.append(f.msb - f.lsb + 1)
 %>\
+   % endif
+
+  // define local fpv variable for multi-reg
+   % if r.get_n_bits(["q", "d"]):
+     % for fpv_name in mreg_fpv_name_list:
+${declare_fpv_var(fpv_name, int(mreg_width_list[loop.index] * mreg_num_regs - 1))}\
+     % endfor
+  for (genvar s = 0; s < ${int(mreg_num_regs)}; s++) begin : gen_${mreg_name}_q
+     % for fpv_name in mreg_fpv_name_list:
+${assign_fpv_var(fpv_name, mreg_dut_path_list[loop.index], int(mreg_width_list[loop.index]))}\
+     % endfor
+  end
+   % endif
+  % endif
+
   % for reg_flat in r.get_regs_flat():
 <%
   reg_name = reg_flat.name
   reg_width = block.addr_width
   reg_offset =  str(reg_width) + "'h" + "%x" % reg_flat.offset
   reg_msb = reg_flat.width - 1
   regwen = reg_flat.regwen
-  reg_access = set()
   reg_wr_mask = 0
   hw_access = reg_flat.fields[0].hwaccess
 %>\
+  // assertions for register: ${reg_name}
     % if regwen:
-<%
-    reg_wen = "i_"+ block.name + ".i_reg_top."+ regwen + "_qs"
-%>\
+<% reg_wen = "dut.u_reg." + regwen + "_qs" %>\
```

Comment:
I believe here I still have to probe the design to get the path for `regwen` register values.
I am probing the `regwen` value to predict the correct write values for locked registers under regwen reg.

Regwen register values only stayed in `reg_top.sv` and does not store inside `reg2hw` or `hw2reg`.
I could not directly probe the value through bind file because some modules might not have regwen registers, some might have more than one regwen regs.

Example of the regwen reg in `reg_top.sv`: https://github.com/lowRISC/opentitan/blob/master/hw/ip/pinmux/rtl/pinmux_reg_top.sv#L648